[Site Creation] Segments Screen

ctarda opened the pull request:
Early pull request, taking the implementation of  #10322 as an excuse to add most, if not all, of the basic scaffolding necessary to build the UI of the new Site Creation flow.

In this PR:
- First implementation of a `WizardStep`. 
- First implementation of the View Controller that would render all the steps headers
- First implementation of the View Controller that renders the Segments Screen, plus cell and datasource
- A few model objects supporting the implementation.
- A builder object, that we'd use to propagate the state of the process (track the data collected from user input/selection)
- The container view controller that we'd use to render the wizard
- A class that puts everything together (create steps, container, set them up for collaboration) and launches the wizard

To have in mind:
- The actual UI of the Segments Screen is (obviously) a draft
- I can not make the `SiteSegmentsDataSource` to run. I assume it will have to be because of the way I am bootstrapping the UI, but I am at that moment when the trees do not let me see the forest.
- I didn't add tests yet. I wanted to get this out as soon as possible. 

To test:
- Make sure the feature flag is active.
- Navigate to the My Sites tab, and tap the plus button in the navigation bar.
- Tap "Enhanced site creation" 
![simulator screen shot - iphone xr - 2018-10-31 at 23 03 09](https://user-images.githubusercontent.com/2722505/47835460-442a7700-dd61-11e8-8245-817c2a96e6f7.png)

stevebaranski commented on WordPress/WordPressTest/SiteSegmentsStepTests.swift:
```diff
@@ -0,0 +1,45 @@
+
+    override func setUp() {
+        super.setUp()
+        step = SiteSegmentsStep(service: MockSiteSegmentsService())
```
I'm seeing the following error when I attempt to run the tests:
```
.../Code/GitHub/WordPress-iOS/WordPress/WordPressTest/SiteSegmentsStepTests.swift:14:33: Missing argument for parameter 'builder' in call
```

It looks like the initializer now includes an additional parameter : `builder`.

stevebaranski commented on WordPress/Classes/ViewRelated/New Group/SiteCreationOutput.swift:
```diff
@@ -0,0 +1,3 @@
+// The final output of the site creation wizard, is an endocable model object that can be submitted to the backend
```
Minor typo : `endocable` -> `encodable`.

stevebaranski commented on WordPress/Classes/ViewRelated/New Group/SiteCreationBuilder.swift:
```diff
@@ -0,0 +1,12 @@
+// Tracks data state shared between Site Creation Wizard Steps. I am not too fond of the name, but it kind of works for now.
```
>  I am not too fond of the name, but it kind of works for now.

`SiteCreator` maybe? The `build` method still underscores that it's a kind-of builder.

stevebaranski commented on WordPress/Classes/ViewRelated/Blog/BlogListViewController.m:
```diff
@@ -882,11 +882,30 @@ - (UIAlertController *)makeAddSiteAlertController
     return addSiteAlertController;
 }
 
+- (void)enhancedSiteCreation
```
@nheagy raises an interesting point. 

Right now we have an `@objc` annotation in `SiteCreationWizardLauncher`. Creating a Swift-based extension would allow us to move the annotation, and it might make it clearer where we are exposing behavior to Objective-C.

stevebaranski commented on WordPress/Classes/ViewRelated/New Group/SiteCreationOutput.swift:
```diff
@@ -0,0 +1,3 @@
+// The final output of the site creation wizard, is an endocable model object that can be submitted to the backend
+struct SiteCreationOutput {
```
The existing implementation uses a struct named `SiteCreationParams` to interface with `SiteCreationService`. There might be opportunities for reuse (or extension) of the service, but that's outside the scope of the PR. I suppose I'm calling it out so we can think about it as we move forward.

stevebaranski commented on WordPress/Classes/ViewRelated/New Group/SiteSegments/SiteSegmentsCell.xib:
```diff
@@ -0,0 +1,62 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
I recognize that we use XIBs & storyboards elsewhere in the app? I like these tools, but I find the XML to be inscrutable, especially during PR reviews.

How would you feel about doing all of our layout programmatically?

stevebaranski commented on WordPress/Classes/ViewRelated/Site Creation/SiteCreationWizard.swift:
```diff
@@ -1,12 +1,20 @@
+        WizardViewController()
+    }()
+
+    lazy var ui: UIViewController = {
```
I like the name `content` above; I wonder if we could rename this property from `ui` to `contentViewController` in the `Wizard` protocol definition?

ctarda replied:
I see your point. `contentViewController` sounds a bit redundant to me though (we'd have the type in the name and the actual type definition). I'll try `content`, just like that, and rename the private one.

stevebaranski commented on WordPress/Classes/ViewRelated/New Group/SiteSegments/SiteSegmentsWizardContent.swift:
```diff
@@ -0,0 +1,51 @@
+    }
+
+    private func handleError(_ error: Error) {
+        print("=== handling error===")
```
How would you feel about adding either a log call via Cocoa Lumberjack, or even `debugPrint`?

ctarda commented:
Thanks for the review, @nheagy and @stevebaranski !

I think I have addressed everything, re-interpreting my own way a couple of your vary valid points 😛

Sorry about the tests. I added the builder at the very last moment, and forgot to run the tests again.

Regarding autolayout, I am more than open to build it programatically, specially considering it's going to be mostly cells, or small views, but I am a little hesitant to do it considering the rest of the project is xib based. If you feel strongly we should go with programatic layout, I say let's do it!

stevebaranski commented:
@ctarda Thanks for indulging the feedback! I have checked out the updates you've applied, and can confirm that the new tests run as expected.

:shipit:
